src/Apps/Order/Components/ShippingAndPaymentReview.tsx

```diff
@@ -43,13 +62,19 @@ export const ShippingAndPaymentReviewFragmentContainer = createFragmentContainer
   ShippingAndPaymentReview,
   graphql`
     fragment ShippingAndPaymentReview_order on Order {
-      fulfillmentType
-      shippingName
-      shippingAddressLine1
-      shippingAddressLine2
-      shippingCity
-      shippingPostalCode
-      shippingRegion
+      requestedFulfillment {
+        ... on Pickup {
+          fulfillmentType
```

Comment:
Rather than query this field, query for `__typename` of the union (I.e. hoist one level up). That way you can always use its value and the generated types will reflect that. It’s also more standard GraphQL usage.

@ashkan18 this field should imo not even be exposed in the schema, only on the input type to work around the lack of union input types.

Reply:
See [here](
https://github.com/artsy/reaction/pull/1193/files/bd97d6d100cec49e6e11ec71dcd185505749ec5b#r21363642)4 for some details about why we want to encourage selection for `__typename` and thus removing these unnecessary fields from the schema completely would be preferable to me.

Reply:
Okay, nearly got this sorted. I think that `fulfillmentType` might _have_ to be included, actually. We need to spread over `Pickup` here in order for Relay to recognize `__typename` might be `Pickup` and not just `Ship`, but in order to spread over a type, it needs to have at least one field. At least that's my working understanding.

Reply:
I took a quick check on master and it seems to work fine when I remove the `Pickup` inline fragment, you just need to make sure that you change the guard to check for `Ship` rather than `Pickup` https://gist.github.com/alloy/1140d14e4b1f487fb207e997b9ec9283

Reply:
I'll take a look – I was getting a type check failure.

Reply:
Ah I see. Here’s a new version of the patch that changes the story to use the sentinel value `%other` which stands for “any other union member than the types being queried for” https://gist.github.com/alloy/1140d14e4b1f487fb207e997b9ec9283#file-shippingandpaymentreview-tsx-diff-L49. (Relay will always include that one by default to be forward compatible for when new union member types are added that existing clients don’t know how to handle yet.)

> I realize that _technically_ this is data might not be needed, but it would also arguably be weird to have requested fulfillment type be one of two things, but developers can only check against one of them.

Not sure I understand. The `Pickup` type doesn’t have any useful data of its own yet, when it does you can query for it. Until then there’s no real point in doing so, it only leads to more data being fetched than would otherwise happen (a field like `__typename` will already implicitly be included by Relay by default on unions, so adding an explicit selection for it doesn’t actually fetch any more data).

Reply:
Ah, that makes sense. Thanks. I'll send a follow-up PR. 

To clarify my other point, it just feels weird to have a type that's "Pickup" or "Ship", but the type-checker only allows it to be compared to "Ship", that's all. It's totally fine, just weird!

Reply:
Opened #1198 and assigned to Justin to spread this around a bit.